File: mrp_app_v7/mrp_app/data/sample.py
"""data/sample.py — reproducible sample data generator for demo/testing."""
from __future__ import annotations

import io
import numpy as np
import pandas as pd
from datetime import date, timedelta


def _date_range(start: date, n: int) -> list[date]:
    return [start + timedelta(days=i) for i in range(n)]


def make_sample_sales(n_history_days: int = 180) -> pd.DataFrame:
    """
    Generate realistic synthetic daily sales data for 5 cafe products.

    Includes:
    - Weekly seasonality (weekend boost)
    - Slight upward trend
    - Gaussian noise
    """
    rng = np.random.default_rng(42)
    today = date.today()
    start = today - timedelta(days=n_history_days)
    dates = _date_range(start, n_history_days)

    products = [
        {"name": "Kopi Susu",    "base": 25, "std": 6},
        {"name": "Matcha Latte", "base": 18, "std": 5},
        {"name": "Croissant",    "base": 40, "std": 10},
        {"name": "Sandwich",     "base": 30, "std": 8},
        {"name": "Smoothie",     "base": 15, "std": 4},
    ]

    rows = []
    for d in dates:
        for p in products:
            dow_mult = 1.30 if d.weekday() >= 5 else 1.0
            trend = 1.0 + 0.0003 * (d - start).days
            qty = max(0, int(rng.normal(p["base"] * dow_mult * trend, p["std"])))
            rows.append({"date": d.isoformat(), "product_name": p["name"], "sales_qty": qty})

    return pd.DataFrame(rows)


def make_sample_bom() -> pd.DataFrame:
    """Return a Bill of Materials table with realistic cafe ingredient ratios."""
    rows = [
        ("Kopi Susu",    "Susu Segar",     0.150),
        ("Kopi Susu",    "Kopi Bubuk",     0.015),
        ("Kopi Susu",    "Gula Pasir",     0.020),
        ("Kopi Susu",    "Es Batu",        0.100),
        ("Matcha Latte", "Susu Segar",     0.180),
        ("Matcha Latte", "Matcha Powder",  0.010),
        ("Matcha Latte", "Gula Pasir",     0.015),
        ("Matcha Latte", "Es Batu",        0.080),
        ("Croissant",    "Tepung Terigu",  0.080),
        ("Croissant",    "Mentega",        0.040),
        ("Croissant",    "Telur",          0.050),
        ("Croissant",    "Gula Pasir",     0.010),
        ("Sandwich",     "Roti Tawar",     0.100),
        ("Sandwich",     "Daging Ayam",    0.080),
        ("Sandwich",     "Sayuran Mix",    0.050),
        ("Smoothie",     "Buah Mix",       0.200),
        ("Smoothie",     "Susu Segar",     0.100),
        ("Smoothie",     "Es Batu",        0.150),
        ("Smoothie",     "Madu",           0.015),
    ]
    return pd.DataFrame(rows, columns=["product_name", "material", "component_qty"])


def sales_to_csv_bytes(df: pd.DataFrame) -> bytes:
    return df.to_csv(index=False).encode("utf-8")


def bom_to_csv_bytes(df: pd.DataFrame) -> bytes:
    return df.to_csv(index=False).encode("utf-8")


def sales_csv_string() -> str:
    return make_sample_sales().to_csv(index=False)


def bom_csv_string() -> str:
    return make_sample_bom().to_csv(index=False)

File: lib/sample.ts
/**
 * lib/sample.ts — generates reproducible sample data for demo.
 */

import type { SalesRow, BomRow } from './types'

function addDays(d: Date, n: number): Date {
  const r = new Date(d); r.setDate(r.getDate() + n); return r
}
function fmt(d: Date): string {
  const y = d.getFullYear()
  const m = String(d.getMonth() + 1).padStart(2, '0')
  const day = String(d.getDate()).padStart(2, '0')
  return `${y}-${m}-${day}`
}

// LCG RNG for reproducibility
function makeLcg(seed: number) {
  let s = seed
  return (mn: number, mx: number) => {
    s = (s * 1664525 + 1013904223) & 0xffffffff
    return mn + ((s >>> 0) / 0xffffffff) * (mx - mn)
  }
}

export function makeSampleSales(nDays = 180): SalesRow[] {
  const rand = makeLcg(42)
  const today = new Date(); today.setHours(0, 0, 0, 0)
  const products = [
    { name: 'Kopi Susu',    base: 25, std: 6 },
    { name: 'Matcha Latte', base: 18, std: 5 },
    { name: 'Croissant',    base: 40, std: 10 },
    { name: 'Sandwich',     base: 30, std: 8 },
    { name: 'Smoothie',     base: 15, std: 4 },
  ]
  const rows: SalesRow[] = []
  for (let i = 0; i < nDays; i++) {
    const d = addDays(today, i - nDays)
    for (const p of products) {
      const dow = (d.getDay() === 0 || d.getDay() === 6) ? 1.3 : 1
      const trend = 1 + 0.0003 * i
      const qty = Math.max(0, Math.round(p.base * dow * trend + rand(-p.std, p.std)))
      rows.push({ date: fmt(d), product_name: p.name, sales_qty: qty })
    }
  }
  return rows
}

export function makeSampleBom(): BomRow[] {
  return [
    { product_name: 'Kopi Susu',    material: 'Susu Segar',     component_qty: 0.15 },
    { product_name: 'Kopi Susu',    material: 'Kopi Bubuk',     component_qty: 0.015 },
    { product_name: 'Kopi Susu',    material: 'Gula Pasir',     component_qty: 0.02 },
    { product_name: 'Kopi Susu',    material: 'Es Batu',        component_qty: 0.1 },
    { product_name: 'Matcha Latte', material: 'Susu Segar',     component_qty: 0.18 },
    { product_name: 'Matcha Latte', material: 'Matcha Powder',  component_qty: 0.01 },
    { product_name: 'Matcha Latte', material: 'Gula Pasir',     component_qty: 0.015 },
    { product_name: 'Matcha Latte', material: 'Es Batu',        component_qty: 0.08 },
    { product_name: 'Croissant',    material: 'Tepung Terigu',  component_qty: 0.08 },
    { product_name: 'Croissant',    material: 'Mentega',        component_qty: 0.04 },
    { product_name: 'Croissant',    material: 'Telur',          component_qty: 0.05 },
    { product_name: 'Croissant',    material: 'Gula Pasir',     component_qty: 0.01 },
    { product_name: 'Sandwich',     material: 'Roti Tawar',     component_qty: 0.1 },
    { product_name: 'Sandwich',     material: 'Daging Ayam',    component_qty: 0.08 },
    { product_name: 'Sandwich',     material: 'Sayuran Mix',    component_qty: 0.05 },
    { product_name: 'Smoothie',     material: 'Buah Mix',       component_qty: 0.2 },
    { product_name: 'Smoothie',     material: 'Susu Segar',     component_qty: 0.1 },
    { product_name: 'Smoothie',     material: 'Es Batu',        component_qty: 0.15 },
    { product_name: 'Smoothie',     material: 'Madu',           component_qty: 0.015 },
  ]
}

export function salesToCsv(rows: SalesRow[]): string {
  return ['date,product_name,sales_qty', ...rows.map(r => `${r.date},${r.product_name},${r.sales_qty}`)].join('\n')
}

export function bomToCsv(rows: BomRow[]): string {
  return ['product_name,material,component_qty', ...rows.map(r => `${r.product_name},${r.material},${r.component_qty}`)].join('\n')
}
